test/SILOptimizer/pound_assert.swift

```diff
@@ -0,0 +1,146 @@
+}
+
+func test_assertionFailure() {
+  #assert(isOne(2)) // expected-error{{assertion failed}}
+  #assert(isOne(2), "2 is not 1") // expected-error{{2 is not 1}}
+}
+
+func test_nonConstant() {
+  #assert(isOne(Int(readLine()!)!)) // expected-error{{#assert condition not constant}}
+  #assert(isOne(Int(readLine()!)!), "input is not 1") // expected-error{{#assert condition not constant}}
+}
+
+// We don't support mutation, so the only loop we can make is infinite.
+// TODO: As soon as we support mutation, add tests with finite loops.
+func infiniteLoop() -> Int {
+  // expected-note @+2 {{condition always evaluates to true}}
+  // expected-note @+1 {{control flow loop found}}
+  while true {}
+  // expected-warning @+1 {{will never be executed}}
+  return 1
+}
+
+func test_infiniteLoop() {
+  // expected-error @+1 {{#assert condition not constant}}
+  #assert(infiniteLoop() == 1)
+}
+
+func recursive(a: Int) -> Int {
+  if a == 0 { return 0 }     // expected-note {{expression is too large to evaluate at compile-time}}
+  return recursive(a: a-1)
+}
+
+func test_recursive() {
+  // expected-error @+1 {{#assert condition not constant}}
+  #assert(recursive(a: 20000) > 42)
+}
+
+func conditional(_ x: Int) -> Int {
+  if x < 0 {
+    return 0
+  } else {
+    return x
+  }
+}
+
+func test_conditional() {
+  #assert(conditional(-5) == 0)
+  #assert(conditional(5) == 5)
+
+  // expected-error @+1 {{assertion failed}}
+  #assert(conditional(-5) == 1)
+  // expected-error @+1 {{assertion failed}}
+  #assert(conditional(5) == 1)
+}
+
+//===----------------------------------------------------------------------===//
+// Top-level evaluation
+//===----------------------------------------------------------------------===//
+
+func test_topLevelEvaluation(topLevelArgument: Int) {
+  let topLevelConst = 1
+  #assert(topLevelConst == 1)
+
+  // TODO: I think that this should get rejected. Otherwise, constness of vars
+  // depends on what you do to them, which is confusing. Perhaps some other
+  // const folder is getting to this before ConstExpr.cpp does?
+  var topLevelVar = 1 // expected-warning {{never mutated}}
+  #assert(topLevelVar == 1)
+
+  var topLevelVarConditionallyMutated = 1
+  if topLevelVarConditionallyMutated < 0 {
+    topLevelVarConditionallyMutated += 1
+  }
+  // expected-error @+2 {{#assert condition not constant}}
+  // expected-note @+1 {{could not fold operation}}
+  #assert(topLevelVarConditionallyMutated == 1)
+
+  // expected-error @+1 {{#assert condition not constant}}
+  #assert(topLevelArgument == 1)
+}
+
+//===----------------------------------------------------------------------===//
+// Integers
+//===----------------------------------------------------------------------===//
+
+func test_trapsAndOverflows() {
+  // The error message below is generated by the traditional constant folder.
+  // expected-error @+1 {{arithmetic operation '124 + 92' (on type 'Int8') results in an overflow}}
+  #assert((124 as Int8) + 92 < 42)
```

Comment:
It seems that the other constant folder is replacing the condition in this case, but leaving the condition alone in the next case. I have added some comments explaining that this is happening, but I haven't investigated why the other constant folder is behaving like that.

Reply:
@marcrasi  I see. Thanks for the clarification. Probably the diagnostic constant propagation, in this case, replaces the addition with a truncated value (besides recording the diagnostics). So the interpreter never sees the addition. Btw, a minor comment is that I think it would be better in these comments to use the term "interpreter"  instead of constant folder for #assert. It may create less confusion if we use the terminology "folder" for something that actually changes the code, whereas the interpreter doesn't.  We may want to eventually rewrite the constant folder using the interpreter, but still I think the folding will be done by the constant folder phase.

Reply:
That sounds like a good rewording. I have updated the comment.